Docs: Add OVN network developer documentation

markylaing opened the pull request:
This adds developer documentation for our OVN network driver. It gives an overview of the components/concepts in OVN that we use and how they are constructed. 

I have tried to keep this very targeted to only things that we need to know as LXD developers, there should be little overlap between OVN documentation or with LXD documentation.

I've added quite a few reviewers as I'm aiming for this to be a) correct, b) understandable to most, c) low-level enough to be useful.

hamistao commented on lxd/network/openvswitch/docs/OVN developer documentation for LXD.md:
```diff
@@ -0,0 +1,125 @@
+These are usually represented in tables in the OVN northbound database.
+
+### Chassis
+A chassis is where traffic physically ingresses or egresses from the virtual network.
```
Maybe `ingresses **to** or egresses from`

tomponline commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+### Logical switch
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
+There are two logical switches per LXD OVN network:
+- The external switch, which is between the logical router and the chassis group (which indirects the to uplink bridge, which may be on another cluster member).
```
"the to uplink"?

markylaing replied:
Hmm yeah I meant uplink bridge here to mean the OVS bridge that sits before the uplink network.

tomponline commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+
+### Port groups
+When an instance is attached via the integration bridge to the internal switch of a LXD OVN network, the logical switch port that is created is added to the port group for that network.
+This is used for ACL configuration so that rules can be applied on the whole address set rather than individual ports.
```
whole address? whole network right?

hamistao replied:
whole address *set*, but probably network is more accurate.

hamistao replied:
whole address **set**, but probably network is more accurate.

markylaing replied:
I got this terminology from the OVN northbound DB schema docs. A port group can be used in a `match` rule for the ACL table, as can address sets. When a port group is created or modified, entries in the `Address_Set` table are automatically populated. So using `Port_Groups` simplifies management of ACLs.

I struggled to put that into words :laughing:

minaelee replied:
I suggest preceding this sentence with a brief description of the phrase "address set" as used in OVN.

tomponline replied:
ah i missed the "set" parse as i skimmed it.

markylaing commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
+There are two logical switches per LXD OVN network:
+- The external switch, which is between the logical router and the chassis group (which indirects the to uplink bridge, which may be on another cluster member).
+- The internal switch, which is between the logical router and the integration bridge. This switch contains DHCP and IP allocation configuration.
```
As far as I can tell in the OVN driver `setup(update bool)` function, IP allocation and DHCP configuration is set on the internal switch. IPv6 advertisements are set on the port. Perhaps I'm interpreting wrong?

markylaing commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+This is by design. It allows OVN to push configuration down to the port level so that packets are handled as quickly as possible.
+
+### Port groups
+When an instance is attached via the integration bridge to the internal switch of a LXD OVN network, the logical switch port that is created is added to the port group for that network.
```
Yep it's two stages but only want to highlight the port group here. Maybe I can re-word it to:

> When a LXD OVN network is created a port group will be created that is namespaced to that network. When instances are connected to the network, logical switch ports are created for them on the internal switch. These logical switch ports are added to the port group for the network.

tomponline replied:
yeah sounds better as nothing to do with integration bridge.
Can you avoid "namespaced" though, as that has connotations for Linux namespaces, can we just say "specific"?

mionaalex reviewed (approved):
@markylaing I cannot judge the correctness of it, but I do find it both understandable and useful, especially with the addition of diagrams. It neatly explains the concepts and how they're connected.

markylaing commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+
+### Logical switch
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
+There are two logical switches per LXD OVN network:
+- The external switch, which is between the logical router and the chassis group (which indirects the to uplink bridge, which may be on another cluster member).
+- The internal switch, which is between the logical router and the integration bridge. This switch contains DHCP and IP allocation configuration.
```
No not quite. There are two logical switches per LXD OVN network. The external switch is used to connect the OVS bridge that is connected to the uplink network. The internal switch is used to connect instances. There is only one internal switch, but there are multiple ports connected to the internal switch, which instances are connected to via the integration bridge.

tomponline replied:
> The external switch is used to connect the OVS bridge that is connected to the uplink network

Not quite.

The external (logical) switch is used to connect the provider OVS bridge to the OVN (logical) router's external port.

The uplink network itself is connected to the provider OVS bridge, so the flow is like this:

uplink network parent interface <-> provider OVS switch <-> external (OVN logical) switch port <-> external (OVN logical) switch <-> external (OVN logical) router port <-> router (OVN logical) <-> internal (OVN logical) router port <-> internal (OVN logical) switch <-> internal (OVN logical) switch ports <-> integration OVS bridge <-> Instance interfaces

simondeziel reviewed (commented):
For some reason, I couldn't see the new content at https://canonical-ubuntu-documentation-library--14556.com.readthedocs.build/lxd/en/14556/contributing/

Thanks for the new (useful) diagrams! I was wondering if we should consider [Mermaid diagrams](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams) (IIRC Gabriel used them in the past). Maybe the tool you used to create the `.png` supports Mermaid ones?

markylaing commented:
> For some reason, I couldn't see the new content at https://canonical-ubuntu-documentation-library--14556.com.readthedocs.build/lxd/en/14556/contributing/

This isn't in the main docs yet. Tom has asked for it to be moved to the LXD docs internals section so I'll move it there and post a link when it's built.

> Thanks for the new (useful) diagrams! I was wondering if we should consider [Mermaid diagrams](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams) (IIRC Gabriel used them in the past). Maybe the tool you used to create the `.png` supports Mermaid ones?

As I'll be moving it to the main docs I don't think this will work with sphinx (though there looks to be a [plugin](https://pypi.org/project/sphinxcontrib-mermaid/) for it). For now I'll change to .svg

simondeziel commented:
> As I'll be moving it to the main docs I don't think this will work with sphinx (though there looks to be a [plugin](https://pypi.org/project/sphinxcontrib-mermaid/) for it). For now I'll change to .svg

Perfect, SVG are nice, thanks!

escabo commented:
My only comment is would it make sense to add our default naming (br-int, lxdovn1) on the drawings?

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+This handles L3 networking and additionally has associated NAT rules and security policies.
+
+### Logical switch
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
```
```suggestion
Logical routers cannot be directly connected to OVS bridges; for this, we use a logical switch.
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+### Logical switch
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
+There are two logical switches per LXD OVN network:
+- The external switch, which is between the logical router and the chassis group (which indirects the to uplink bridge, which may be on another cluster member).
```
```suggestion
- The external switch, which is between the logical router and the chassis group. This switch indirects to the OVS bridge that sits before the uplink network. The OVS bridge can be on a different cluster member.
```

minaelee replied:
Also - "...can be on a different cluster member" is unclear to me. What is it different from?

markylaing replied:
I've reworded this section.

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+- The internal switch, which is between the logical router and the integration bridge. This switch contains DHCP and IP allocation configuration.
+
+### Logical switch/router ports
+In OVN, when you create a router or switch it doesn't initially have any ports.
```
```suggestion
When you create a router or switch in OVN, it doesn't initially have any ports.
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+![Physical uplink diagram](ovn-uplink-physical.png)
+
+> Note: When using a physical network as an uplink for OVN, any IP addresses on the parent interface will become defunct.
+> It is a requirement that the parent network should have no network configuration.
```
```suggestion
> The parent network must not have any assigned IP addresses .
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+Any necessary resources for the uplink are created on network start. Networks are usually started immediately after they are created.
+
+### Bridge (OpenVSwitch)
+A user may separately configure an OpenVSwitch bridge.
```
```suggestion
A user can separately configure an OpenVSwitch bridge.
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+
+### Bridge (native)
+A native Linux bridge can also be used.
+In this case we perform the same steps as in the physical network and additionally configure a veth pair.
```
```suggestion
In this case, we perform the same steps as in the physical network and additionally configure a veth pair.
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+### Bridge (native)
+A native Linux bridge can also be used.
+In this case we perform the same steps as in the physical network and additionally configure a veth pair.
+The veth pair is used so that the bridge can still be used for other purposes (i.e. it will maintain it's configuration).
```
```suggestion
The veth pair is used so that the bridge can still be used for other purposes (since the bridge maintains its configuration).
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+![LXD OVN network diagram](ovn-network.png)
+
+> Note: This diagram does not show cross-cluster networks.
+> This conceptual diagram should look this same on all cluster members.
```
```suggestion
> This conceptual diagram should look the same on all cluster members.
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+
+> Note: This diagram does not show cross-cluster networks.
+> This conceptual diagram should look this same on all cluster members.
+> If the chassis group prioritises another chassis for the uplink, the traffic will be routed to the other member by means that we don't need to worry about.
```
```suggestion
> If the chassis group prioritizes another chassis for the uplink, the traffic is routed to that chassis by means beyond the scope of this document.
```

minaelee replied:
(or just stop the sentence at "... routed to that chassis."

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+## Integration bridge
+The cluster setting `network.ovn.integration_bridge` must contain the name of an OVS bridge that is used to connect instances to an OVN network via a NIC device. This OVS bridge must be pre-configured on all cluster members with the same name. 
+
+Connectivity to the integration bridge differs between containers and virtual machines. See below:
```
```suggestion
Connectivity to the integration bridge differs between containers and virtual machines:
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+In LXD, there will usually be one chassis per cluster member.
+If LXD is configured to use OVN networking, then all members *can* be used as OVN chassis.
+
+> If any cluster members have the role `ovn-chassis`, only those members will be represented as chassis in the chassis group table (see below).
```
```suggestion
> If any cluster members have the role `ovn-chassis`, only those members are represented as chassis in the chassis group table (see below).
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+If LXD is configured to use OVN networking, then all members *can* be used as OVN chassis.
+
+> If any cluster members have the role `ovn-chassis`, only those members will be represented as chassis in the chassis group table (see below).
+> If no members have the role, all cluster members will be added to the chassis group.
```
```suggestion
> If no members have the role, all cluster members are added to the chassis group.
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+
+### Logical switch/router ports
+In OVN, when you create a router or switch it doesn't initially have any ports.
+You need to create ports and then link them.
```
Is this true in LXD's implementation of OVN? That is, when you create an OVN network in LXD, does LXD configure the ports automatically or do they need to be manually configured?

tomponline replied:
These docs are not about LXD's OVN networks per-se, its about how LXD OVN networks are built using OVN. So its important to note here that this isn't necessarily for end users (unless they are interested in how LXD configures OVN).

It is true that OVN's routers dont have ports until they are added, and this is why LXD's OVN implementation adds them on the users behalf as part of its opinionated setup.

Shall we move this into the https://documentation.ubuntu.com/lxd/en/latest/internals/ section of the docs btw?

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+This is used for ACL configuration so that rules can be applied on the whole address set rather than individual ports.
+
+## OVN Uplink
+There are three kinds of uplink configurations.
```
```suggestion
There are three types of uplink configurations.
```

tomponline replied:
My only word of caution here is that using "types" suggests that LXD will have a documented (and displayed) understanding of these "types" of configurations (like other sorts of "types" we have), where as this is not the case. LXD will detect the environment of the uplink network and do what is needed to connect it to the OVN space.

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+### Bridge (OpenVSwitch)
+A user may separately configure an OpenVSwitch bridge.
+An OVN network can be created with `network` set to the OVS bridge.
+LXD will configure a bridge mapping on it to connect the OVN network:
```
```suggestion
LXD configures a bridge mapping on the OVS bridge to connect the OVN network:
```

minaelee commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+> It is a requirement that the parent network should have no network configuration.
+
+### Bridge (native)
+A native Linux bridge can also be used.
```
```suggestion
A native Linux bridge can be used.
```

minaelee reviewed (changes requested):
I also cannot comment on the technical accuracy, but it seems well put together overall. I've added some minor feedback regarding clarity/consistency/grammar.

tomponline commented:
Asked @fnordahl for his thoughts too

markylaing commented on lxd/network/openvswitch/docs/developer-documentation.md:
```diff
@@ -0,0 +1,127 @@
+
+### OVS (OpenVSwitch) Bridge
+OVS bridges are used to connect virtual networks to physical ones and vice-versa.
+In LXD, if OpenVSwitch APIs are invoked directly, that means changes are being applied on the same physical cluster member.
```
I'll perhaps change this to be `the daemon host`

markylaing commented:
I have updated with:
- Suggested changes from @minaelee and @tomponline (thanks)
- Moved to "internals" section of main LXD docs.
- Images are now svg.

markylaing commented:
I'm not certain why the documentation tests failed. Markdown lint is complaining about a file I haven't touched: https://github.com/canonical/lxd/actions/runs/12138073445/job/33842728868?pr=14556#step:5:16

simondeziel commented:
> I'm not certain why the documentation tests failed. Markdown lint is complaining about a file I haven't touched: https://github.com/canonical/lxd/actions/runs/12138073445/job/33842728868?pr=14556#step:5:16

Not entirely clear but I suspect that this is related to some docs including other pages causing offsets to change in unrelated files. I had to workaround it with https://github.com/canonical/lxd/pull/14562/commits/52f86bccce0c99bb737c3b2efcf6512e658231f4

simondeziel commented on doc/reference/ovn-internals.md:
````diff
@@ -0,0 +1,168 @@
+```{note}
+This diagram does not show cross-cluster networks.
+This conceptual diagram should look the same on all cluster members.
+If the chassis group prioritizes another chassis for the uplink, the traffic is routed that chassis.
````
```suggestion
If the chassis group prioritizes another chassis for the uplink, the traffic is routed through that chassis.
```

simondeziel commented on doc/reference/ovn-internals.md:
````diff
@@ -0,0 +1,168 @@
+- Containers use a `veth` pair (similar to connecting to a native bridge uplink network).
+
+    ```{figure} /images/ovn/ovn-integration-bridge-container.svg
+    Integration bridge connectivity with containers
````
In that diagram, between the `Veth pair`, is the VLAN filter also applied lik in the native bridge uplink network case?

markylaing replied:
I couldn't see any evidence of that in the code. What are your thoughts @tomponline?

simondeziel reviewed (commented):
Nicely written and the (SVG!) diagrams make this really clear. Thanks!

markylaing commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+These are usually represented in tables in the OVN northbound database.
+
+### Chassis
+A chassis is where traffic physically ingresses to or egresses from the virtual network.
```
Yeah I think that's clearer, will change.

markylaing commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+
+- The external switch, which connects via logical switch port to a port on the logical router and to the provider OVS switch.
+- The internal switch, which connects via logical switch port to a port on the logical router and to the integration bridge.
+  This switch contains DHCP and IP allocation configuration.
```
As far as I can tell, this is where it is logically associated in the OVN northbound database. I'm reading up on this now and will update you.

markylaing replied:
I'm still not 100% certain on this. As far as I can tell, we are creating an entry in the [`DHCP_Options`](https://manpages.ubuntu.com/manpages/noble/en/man5/ovn-nb.5.html#dhcp_options%20table) table that corresponds to the internal switch and subnet.

By "corresponds to" I mean that we use the internal switch name as an `external_id` to reference that entry in the table. 

We do associate that `DHCP_Options` entry with the logical router for the network. But I can't say that the DHCP daemon feature is implemented by the router itself (I guess that's OVN internals we don't need to worry about).

simondeziel replied:
Indeed, that's probably OVN internal details. Thanks!

hamistao commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+
+It is important to note that a `physical` managed network's `parent` interface can be any of these types, and that for a `managed` bridge network the parent interface can be either types of bridges.
+
+### Bridge (OpenVSwitch)
```
Just a reminder, there are some `OpenVSwitch` on the diagrams as well

hamistao commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+- OVS Bridge.
+- Physical interface (or bond).
+
+It is important to note that a `physical` managed network's `parent` interface can be any of these types, and that for a `managed` bridge network the parent interface can be either types of bridges.
```
```suggestion
It is important to note that a `physical` managed network's `parent` interface can be any of these types, and that for a managed `bridge` network the `parent` interface can be either types of bridges.
```

Also a couple of quick questions. In a bridge network, does the chassis group only contain one chassis? And do we include the per-member `parent` interfaces as part of each OVN chassis?

Those are more for my own understanding, but if you think there are relevant differences on how we operate on chassis in each type of network feel free to add something in.

markylaing replied:
I'm not sure I understand the question.

If LXD is running in standalone mode then there is only one chassis (the host) in the chassis group. If LXD is clustered then either all members are in the chassis group, OR only members with the `ovn-chassis` role are in the chassis group.

If there are multiple OVN networks, all OVN networks will contain the same cluster members in their chassis group. However, the chassis group for each network will assign different priorities to the chassis. This keeps uplink traffic distributed amongst the host machines.

Conceptually, the `parent` network is not something that is passed into OVN. We use an OVS bridge to get from the physical network to the virtual one and vice versa.

hamistao reviewed (commented):
I don't think I have anything else to add on top of the other reviews. Very clear, complete and well written, thanks!

kadinsayani commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+# LXD OVN implementation
```
I think `(ovn-internals)=` needs to come before the heading for doc builds. Just something I noticed while working on https://github.com/canonical/lxd/pull/14574 :)

markylaing replied:
This is required for external linking but there are no links to this page yet.

kadinsayani replied:
Ah I see, thanks.

kadinsayani commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+
+For more detailed documentation on OVN itself, please see:
+
+- [Overview of OVN and SDNs](https://ubuntu.com/blog/data-centre-networking-what-is-ovn).
```
Perhaps these links can go in the "relevant links" section?

markylaing replied:
Happy to add them there as well but I'd like to keep them prominent so that they are "required reading", contextually.

markylaing commented on doc/reference/ovn-internals.md:
```diff
@@ -0,0 +1,168 @@
+
+- Linux native bridge.
+- OVS Bridge.
+- Physical interface (or bond).
```
I've added `vlan`.

tomponline commented:
@minaelee please can you do another review?
